nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedTransformRecord.java

```diff
@@ -0,0 +1,398 @@
+
+        final AtomicLong recordIndex = new AtomicLong(0L);
+        final AtomicLong dropCount = new AtomicLong(0L);
+
+        try {
+            final Map<String, String> attributesToAdd = new HashMap<>();
+
+            // Read each record, transform it, and write out the transformed version
+            session.write(flowFile, (in, out) -> {
+                try (final RecordReader reader = readerFactory.createRecordReader(flowFile, in, getLogger());
+                     final RecordSetWriter writer = writerFactory.createWriter(getLogger(), reader.getSchema(), out, flowFile)) {
+
+                    writer.beginRecordSet();
+
+                    Record inputRecord;
+                    while ((inputRecord = reader.nextRecord()) != null) {
+                        final long index = recordIndex.get();
+
+                        // Evaluate the script against the Record
+                        final Object returnValue = evaluator.evaluate(inputRecord, index);
+
+                        recordIndex.getAndIncrement();
+
+                        // If a null value was returned, drop the Record
+                        if (returnValue == null) {
+                            getLogger().trace("Script returned null for Record {} [{}] so will drop Record from {}", new Object[]{index, inputRecord, flowFile});
+                            dropCount.getAndIncrement();
+                            continue;
+                        }
+
+                        // If a single Record was returned, write it out
+                        if (returnValue instanceof Record) {
+                            final Record transformedRecord = (Record) returnValue;
+                            getLogger().trace("Successfully transformed Record {} from {} to {} for {}", new Object[] {index, inputRecord, transformedRecord, flowFile});
+                            writer.write(transformedRecord);
+                            continue;
+                        }
+
+                        // If a Collection was returned, ensure that every element in the collection is a Record and write them out
+                        if (returnValue instanceof Collection) {
+                            final Collection<?> collection = (Collection<?>) returnValue;
+                            getLogger().trace("Successfully transformed Record {} from {} to {} for {}", new Object[] {index, inputRecord, collection, flowFile});
+
+                            for (final Object element : collection) {
+                                if (!(element instanceof Record)) {
+                                    throw new RuntimeException("Evaluated script against Record number " + index + " of " + flowFile
+                                        + " but instead of returning a Record or Collection of Records, script returned a Collection of values, "
+                                        + "at least one of which was not a Record but instead was: " + returnValue);
```

Comment:
```ot a Record but instead was: " + returnValue```

could be

```ot a Record but instead was: " + element.getClass().getName()```

to better show the specific bad value type.

Reply:
I think which one (between the object toString() vs. class name) is more informative depends on the particular object. The default implementation of toString() includes the class name, but if that is overridden, the toString() *should* be more informative.